File: java/src/main/java/com/leecampbell/cod/domain/model/Loan.java
package com.leecampbell.cod.domain.model;

import com.leecampbell.cod.domain.contracts.*;
import java.math.BigDecimal;
import java.time.OffsetDateTime;
import java.util.UUID;

public final class Loan extends AggregateRoot {
    private static final BigDecimal MIN_AMOUNT = new BigDecimal(50);
    private static final BigDecimal MAX_AMOUNT = new BigDecimal(2000);

    private OffsetDateTime createdOn;
    private BigDecimal loanAmount;
    private BigDecimal balance;
    private BankAccount bankAccount;
    private OffsetDateTime disbursedOn;

    public Loan(UUID loanId) {
        super(loanId);
    }

    public void Create(CreateLoanCommand command) {

        if (version() > 0)
            throw new UnsupportedOperationException("Loan already created.");
        if (command.amount().compareTo(MIN_AMOUNT) < 0 || MAX_AMOUNT.compareTo(command.amount()) < 0)
            throw new UnsupportedOperationException("Only loan amounts between $50.00 and $2000.00 are supported.");
        if (!isUnder2Years(command.term()))
            throw new UnsupportedOperationException("Only loan terms up to 2 years are supported.");

        AddEvent(new LoanCreatedEvent(command.createdOn(), command.term(), command.paymentPlan(), command.amount()));
        AddEvent(new LoanCustomerContactChangedEvent(command.customerContact().name(),
                command.customerContact().preferredPhoneNumber(), command.customerContact().alternatePhoneNumber(),
                command.customerContact().postalAddress()));
        AddEvent(new LoanBankAccountChangedEvent(command.bankAccount().bsb(), command.bankAccount().accountNumber()));
    }

    public void DisburseFunds(DisburseLoanFundsCommand command) {
        if (disbursedOn != null)
            throw new UnsupportedOperationException("Funds are already disbursed.");

        BankAccount disburseToAccount = new BankAccount(bankAccount.bsb(), bankAccount.accountNumber());
        AddEvent(new LoanDisbursedFundsEvent(command.transactionDate(), loanAmount.negate(), disburseToAccount));
    }

    public void TakePayment(TakePaymentCommand command) {
        if (createdOn == null)
            throw new UnsupportedOperationException("Take payment attempted on an uncreated loan.");
        if (command.transactionDate().compareTo(createdOn) < 0)
            throw new UnsupportedOperationException("Transaction date can not be prior to loan creation.");
        if (command.transactionAmount().compareTo(BigDecimal.ZERO) <= 0)
            throw new UnsupportedOperationException("Transaction amount must be positive.");

        AddEvent(new PaymentTakenEvent(UUID.randomUUID().toString(), command.transactionDate(),
                command.transactionAmount()));
        if (balance.compareTo(BigDecimal.ZERO) >= 0)
            AddEvent(new LoanSettledEvent(command.transactionDate()));
        if (balance.compareTo(BigDecimal.ZERO) > 0)
            AddEvent(new LoanOverPaidEvent(command.transactionDate(), balance));
    }
    //public void ReverseDisbursement(ReverseDisbursement command) {
    //}
    //public void ReversePayment(ReversePaymentCommand command) {
    //}
    //public void RefundPayment(RefundPaymentCommand command) {
    //}
    //public void PaymentMissed(?? command) {
    //}
    //public void SetRepaymentMethod(SetRepaymentMethodCommand command) {
    //}
    //public void FailLoan(FailLoanCommand command) {
    //}
    //public void CancelLoan(CancelLoanCommand command) {
    //}

    private void handle(LoanCreatedEvent e) {
        createdOn = e.createdOn();
        loanAmount = e.amount();
        balance = BigDecimal.ZERO; //Balance is zero until we disburse the money.
    }

    private void handle(LoanDisbursedFundsEvent e) {
        disbursedOn = e.transactionDate();
        balance = balance.add(e.amount());
    }

    private void handle(LoanCustomerContactChangedEvent e) {
    }

    private void handle(LoanBankAccountChangedEvent e) {
        this.bankAccount = e.bankAccount();
    }

    private void handle(PaymentTakenEvent e) {
        balance = balance.add(e.amount());
    }

    private void handle(LoanSettledEvent e) {
    }

    private void handle(LoanOverPaidEvent e) {
    }

    private static boolean isUnder2Years(Duration duration) {
        switch (duration.Unit()) {
        case Day:
            return duration.length() < 365 * 2;
        case Week:
            return duration.length() < 52 * 2;
        case Month:
            return duration.length() < 12 * 2;
        default:
            throw new IllegalArgumentException("Unsupported duration");
        }
    }
}

File: CSharp/Yow.CoD.Finance.Domain/Model/Loan.cs
using System;
using Yow.CoD.Finance.Domain.Contracts;

namespace Yow.CoD.Finance.Domain.Model
{
    //Make focus on Taking Payment
    //  Reduces balance
    //  May re-calculate payment schedule
    //  May transition the state of the loan (paid, overpaid, out arrears and back into active)

    public class Loan : AggregateRoot
    {
        private BankAccount _bankAccount;
        private DateTimeOffset _createdOn;
        private DateTimeOffset _disbursedOn;
        private decimal _loanAmount;
        private decimal _balance;

        public Loan(Guid loanId) : base(loanId)
        {
            RegisterHandler<LoanCreatedEvent>(Handle);
            RegisterHandler<LoanDisbursedFundsEvent>(Handle);
            RegisterHandler<LoanCustomerContactChangedEvent>(Handle);
            RegisterHandler<LoanBankAccountChangedEvent>(Handle);
            RegisterHandler<PaymentTakenEvent>(Handle);
            RegisterHandler<LoanSettledEvent>(Handle);
            RegisterHandler<LoanOverPaidEvent>(Handle);
        }

        public void Create(CreateLoanCommand command)
        {
            if (Version > 0) throw new InvalidOperationException("Loan already created.");
            if (command.Amount < 50m || 2000 < command.Amount) throw new InvalidOperationException("Only loan amounts between $50.00 and $2000.00 are supported.");
            if (!IsUnder2Years(command.Term)) throw new InvalidOperationException("Only loan terms up to 2 years are supported.");

            AddEvent(new LoanCreatedEvent(command.CreatedOn, command.Amount, command.Term, command.PaymentPlan));
            AddEvent(new LoanCustomerContactChangedEvent(command.CustomerContact.Name, command.CustomerContact.PreferredPhoneNumber, command.CustomerContact.AlternatePhoneNumber, command.CustomerContact.PostalAddress));
            AddEvent(new LoanBankAccountChangedEvent(command.BankAccount.Bsb, command.BankAccount.AccountNumber));
        }

        public void DisburseFunds(DisburseLoanFundsCommand command)
        {
            if (_disbursedOn != default) throw new InvalidOperationException("Funds are already disbursed.");

            var disburseToAccount = new Contracts.BankAccount(_bankAccount.Bsb, _bankAccount.AccountNumber);
            AddEvent(new LoanDisbursedFundsEvent(command.TransactionDate, -_loanAmount, disburseToAccount));
        }

        public void TakePayment(TakePaymentCommand command)
        {
            if (command.TransactionDateTime < _createdOn)
                throw new ArgumentException("Transaction date can not be prior to loan creation", nameof(command));
            if (command.Amount <= decimal.Zero)
                throw new ArgumentException("Transaction amount must be positive", nameof(command));

            AddEvent(new PaymentTakenEvent(Guid.NewGuid().ToString(), command.TransactionDateTime, command.Amount));
            if (_balance >= decimal.Zero)
                AddEvent(new LoanSettledEvent(command.TransactionDateTime));
            if (_balance > decimal.Zero)
                AddEvent(new LoanOverPaidEvent(command.TransactionDateTime, _balance));
        }



        //public void ReverseDisbursement(ReverseDisbursement command)
        //{
        //}
        //public void ReversePayment(ReversePaymentCommand command)
        //{
        //}
        //public void RefundPayment(RefundPaymentCommand command)
        //{
        //}
        //public void PaymentMissed(?? command)
        //{
        //}
        //public void SetRepaymentMethod(SetRepaymentMethodCommand command)
        //{
        //}
        //public void FailLoan(FailLoanCommand command)
        //{
        //}
        //public void CancelLoan(CancelLoanCommand command)
        //{
        //}

        private void Handle(LoanCreatedEvent e)
        {
            _createdOn = e.CreatedOn;
            _loanAmount = e.Amount;
            _balance = decimal.Zero; //Balance is zero until we disburse the money.
        }

        private void Handle(LoanDisbursedFundsEvent e)
        {
            _disbursedOn = e.TransactionDate;
            _balance += e.Amount;
        }

        private void Handle(LoanCustomerContactChangedEvent e)
        {
        }

        private void Handle(LoanBankAccountChangedEvent e)
        {
            _bankAccount = new BankAccount(e.BankAccount.Bsb, e.BankAccount.AccountNumber);
        }

        private void Handle(PaymentTakenEvent e)
        {
            _balance += e.Amount;
        }

        private void Handle(LoanSettledEvent e)
        {
        }

        private void Handle(LoanOverPaidEvent e)
        {
        }

        private static bool IsUnder2Years(Duration duration)
        {
            switch (duration.Unit)
            {
                case DurationUnit.Day:
                    return duration.Length < 365 * 2;
                case DurationUnit.Week:
                    return duration.Length < 52 * 2;
                case DurationUnit.Month:
                    return duration.Length < 12 * 2;
                default:
                    throw new ArgumentOutOfRangeException();
            }
        }
    }
}